src/full/Agda/TypeChecking/Rules/PatternSynDef.hs

```diff
@@ -2,42 +2,151 @@
 
 module Agda.TypeChecking.Rules.PatternSynDef where
 
+import Data.Maybe
 import Data.Void
 
 import Agda.Syntax.Common
+import Agda.Syntax.Info
+
 import qualified Agda.Syntax.Abstract as A
 import qualified Agda.Syntax.Abstract.Views as A
-import Agda.Syntax.Info
+import Agda.Syntax.Internal
 
 import Agda.TypeChecking.Monad
+import Agda.TypeChecking.Reduce
+import Agda.TypeChecking.Conversion
+import Agda.TypeChecking.Datatypes
+import Agda.TypeChecking.Implicit
+
+import Agda.TypeChecking.Rules.LHS ( disambiguateConstructor )
+import Agda.TypeChecking.Rules.Term
+
+import Agda.Utils.Monad
+import Agda.Utils.Impossible
+
 
-checkPatternSynDef :: QName -> Maybe (A.DefInfo, A.Expr) -> [Arg BindName] ->
+checkPatternSynDef :: QName -> Maybe (A.DefInfo, A.Expr) -> [Arg A.BindName] ->
                       A.Pattern' Void -> TCM ()
-checkPatternSynDef name Nothing as rhs = ?
-checkPatternSynDef name (Just (i, e)) as rhs = ?
+checkPatternSynDef name Nothing as rhs = undefined
+checkPatternSynDef name (Just (i, e)) as rhs = do
+  ty <- isType_ e
+  checkTyped name i ty as rhs
 
-checkTyped :: QName -> A.DefInfo -> Type -> [Arg BindName] ->
+checkTyped :: QName -> A.DefInfo -> Type -> [Arg A.BindName] ->
               A.Pattern' Void -> TCM ()
-checkTyped name i ty as rhs = do
-  TelV tel bty <- telViewUpTo (length as) ty
-  unless (size tel == length as) $ genericError "pattern syn telescope not right"
-
--- telViewUpToArgs :: [Arg BindName] -> Type -> m TelView
--- telViewUpToArgs [] t = return $ TelV EmptyTel t
--- telViewUpToArgs (Arg i n : as) t = do
---   t <- reduce t
---   case argInfoHiding a, unEl t of
---     NotHidden, Pi ta tb | NotHidden == getHiding ta ->
---       absV ?
---     _ -> return $ TelV EmptyTel t
---   where
---     absV a x (TelV tel t) = TelV (ExtendTel a (Abs x tel)) t
-
-checkLHS :: QName -> Type -> [Arg BindName] -> (Type -> TCM a) -> TCM a
-checkLHS name ty [] k = k ty
-checkLHS name ty (Arg i n : as) k = do
+checkTyped name i sig as rhs =
+  -- 1. Check left hand side
+  checkLHS sig as $ \ target -> do
+  -- 2. Sanity check: pattern syn targets data or record type
+    target <- reduce target
+    case unEl target of
+      Def tyCon _ -> do
+        Just{} <- isDataOrRecordType tyCon
+        pure ()
+      _ -> genericError "Expected a data or record type as the target of the pattern synonym"
+  -- 3. Check right hand side
+    checkRHS target rhs $ \ rhs ->
+  -- 4. Register "constructor"
+      undefined
+
+checkLHS :: Type -> [Arg A.BindName] -> ({- Telescope -> -} Type -> TCM a) -> TCM a
```

Comment:
Is it right to pattern-match on the list of arguments first? What if the type ends with an implicit Pi, like:
```agda
pattern
  ze : ∀ {n} → Fin (suc n)
  ze = zero
```

Reply:
:thinking: